File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/WorldStep/RuleGroup/PropertyRule/Entry/PropertyThreshold/PropertyThreshold.java
package ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.PropertyRule.Entry.PropertyThreshold;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.logging.LogScope.logScope;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class PropertyThreshold implements  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static String nodeName = "property-threshold";
    public static PropertyThreshold fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        var instance = new PropertyThreshold();
        if(Objects.nonNull(parent)) {
          instance.parentNode(parent);
        }
        instance.rawNode(rawNode);
        instance.deserialize(rawNode);
        return logger.logReturn(instance);
      }

    }
    public static PropertyThreshold fromRawNode(RawNode rawNode) {
      try (var logger = logScope()) {
        var instance = fromRawNode(rawNode, null);
        return logger.logReturn(instance);
      }
    }
    public static Optional<PropertyThreshold> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        try(var logger = logScope()) {
          return logger.logReturn(rawNode.map(o -> PropertyThreshold.fromRawNode(o, parent)));
        }

    }
    public static List<PropertyThreshold> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        List<PropertyThreshold> returnList = Optional.ofNullable(rawNodeList)
            .orElse(List.of())
            .stream()
            .map(o -> PropertyThreshold.fromRawNode(o, parent))
            .collect(Collectors.toList());
        return logger.logReturn(returnList);
      }
    }

    public String classTypeId() {
      return ".world_step.rule_group.property_rule.entry.property-threshold";
    }

    //Attributes

    private String name;
    @Builder.Default
    private Optional<Integer> minValueInclusive = Optional.empty();
    @Builder.Default
    private Optional<Integer> maxValueInclusive = Optional.empty();

    //Children elements

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private RawNode rawNode = new RawNode();

    public RawNode rawNode() {
      return rawNode;
    }
    public void rawNode(RawNode rawNode) {
      this.rawNode = rawNode;
    }

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode() {
      return parentNode;
    }

    @Builder.Default
    private List<Consumer<List<Object>>> onChangeList = new ArrayList<>();

    public String nodeName() {
      return "property-threshold";
    }

    public void notifyChange(List<Object> list) {
      try (var logger = logScope()) {
        list.addLast(this);
        logger.log("Notify change for", this.buildPath());
        onChangeList.forEach(consumer -> consumer.accept(list));
        parentNode.ifPresent(linkedNode -> linkedNode.notifyChange(list));
      }
    }

    public void parentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode = Optional.of(linkedNode);
      notifyChange();
    }

    public Optional<ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.PropertyRule.Entry.Entry> parentAsEntry() {
      return parentNode.flatMap(node -> {
        if (node instanceof ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.PropertyRule.Entry.Entry casted){
          return Optional.of(casted);
        }
        return Optional.empty();
      });
    }

    public void removeChild(Object object) {
    }

    public int buildIndexForChild(Object object) {
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(Consumer<List<Object>> onChange) {
      try (var logger = logScope()) {
        onChangeList.add(onChange);
        return logger.logReturn(() -> onChangeList.remove(onChange));
      }
    }

    public void deserialize (RawNode rawNode) {
      try (var logger = logScope()) {
        this.rawNode = rawNode;
        // Godot.GD.Print("Deserializing property-threshold");

        try (var innerLogger = logScope("attributes")) {
          //Deserialize attributes
          innerLogger.log("name");
          this.name = rawNode.getAttributeRequired("name");
          innerLogger.log("min-value-inclusive");
          this.minValueInclusive = rawNode.getAttributeInt("min-value-inclusive");
          innerLogger.log("max-value-inclusive");
          this.maxValueInclusive = rawNode.getAttributeInt("max-value-inclusive");
        }
        try (var innerLogger = logScope("children")) {
          //Deserialize children
        }
      } catch (Exception e) {
        throw new RuntimeException("Deserialization failed for: " + this.buildPath(), e);
      }

    }

    public RawNode serializeIntoRawNode()
    {
      try (var logger = logScope()) {
        rawNode.setTag("property-threshold");
        try (var innerLogger = logScope("attributes")) {
          //Serialize attributes
          innerLogger.log("name");
          rawNode.setAttribute("name", this.name);
          innerLogger.log("min-value-inclusive");
          this.minValueInclusive.ifPresent(o -> rawNode.setAttribute("min-value-inclusive", o));
          innerLogger.log("max-value-inclusive");
          this.maxValueInclusive.ifPresent(o -> rawNode.setAttribute("max-value-inclusive", o));
        }
        try (var innerLogger = logScope("children")) {

          //Serialize children
          return rawNode;
        }
      }
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing property-threshold");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }

    public String getName()
    {
      return this.name;
    }
    public PropertyThreshold setName(String value)
    {
      this.name = value;
      notifyChange();
      return this;
    }
    public Optional<Integer> getMinValueInclusive()
    {
      return this.minValueInclusive;
    }
    public PropertyThreshold setMinValueInclusive(Optional<Integer> value)
    {
      this.minValueInclusive = value;
      notifyChange();
      return this;
    }
    public Optional<Integer> getMaxValueInclusive()
    {
      return this.maxValueInclusive;
    }
    public PropertyThreshold setMaxValueInclusive(Optional<Integer> value)
    {
      this.maxValueInclusive = value;
      notifyChange();
      return this;
    }

    public ro.anud.xml_xsd.implementation.util.LinkedNode deserializeAtPath(String xpath, RawNode rawNode) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }

        deserialize(rawNode);
        return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> getNodeAtPath(String xpath) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }
        return Optional.of(this);
    }
  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "value": {},
        "isSingle": false,
        "isNullable": true,
        "attributes": {
          "metaType": "object",
          "value": {
            "name": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "min-value-inclusive": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": true
            },
            "max-value-inclusive": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": true
            }
          }
        }
      },
      "name": "property-threshold"
    }
  */

File: gui-client/world_step/world_step/rule_group/property_rule/entry/property_threshold/property-threshold.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Nworld_step.Nrule_group.Nproperty_rule.Nentry.Nproperty_threshold {}
namespace XSD {
}
namespace XSD.Nworld_step.Nrule_group.Nproperty_rule.Nentry {
  public class property_threshold : XSD.ILinkedNode  {

    public static string ClassTypeId = ".world_step.rule_group.property_rule.entry.property-threshold";
    public static string TagName = "property-threshold";

    public string NodeName {get =>"property-threshold";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<property_threshold>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes
    private System.String _name;
    public System.String name { get => _name; set => _name = value; }
    private System.Int32? _min_value_inclusive;
    public System.Int32? min_value_inclusive { get => _min_value_inclusive; set => _min_value_inclusive = value; }
    private System.Int32? _max_value_inclusive;
    public System.Int32? max_value_inclusive { get => _max_value_inclusive; set => _max_value_inclusive = value; }

    //Children elements
    public property_threshold()
    {
    }

    public property_threshold(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public property_threshold(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
      if(name == "name")
      {
        Set_name(value);
      }
      if(name == "min_value_inclusive")
      {
        Set_min_value_inclusive(value?.ToInt() ?? 0);
      }
      if(name == "max_value_inclusive")
      {
        Set_max_value_inclusive(value?.ToInt() ?? 0);
      }
    }

    public void SetChild(dynamic linkedNode)
    {
    }

    public void ClearChild(dynamic linkedNode)
    {
    }

    public Action OnSelfChange(Action<property_threshold> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing property-threshold");
      //Deserialize arguments
      if(rawNode.attributes.ContainsKey("name"))
      {
        var attribute_name = rawNode.attributes["name"];
        this.name = rawNode.attributes["name"];
      }
      if(rawNode.attributes.ContainsKey("min-value-inclusive"))
      {
        var attribute_min_value_inclusive = rawNode.attributes["min-value-inclusive"];
        this.min_value_inclusive = attribute_min_value_inclusive?.ToInt();
      }
      if(rawNode.attributes.ContainsKey("max-value-inclusive"))
      {
        var attribute_max_value_inclusive = rawNode.attributes["max-value-inclusive"];
        this.max_value_inclusive = attribute_max_value_inclusive?.ToInt();
      }

      //Deserialize children
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments
      if(this.name != null)
      {
        rawNode.attributes["name"] = this.name.ToString();
      }
      if(this.min_value_inclusive != null)
      {
        rawNode.attributes["min-value-inclusive"] = this.min_value_inclusive?.ToString();
      }
      if(this.max_value_inclusive != null)
      {
        rawNode.attributes["max-value-inclusive"] = this.max_value_inclusive?.ToString();
      }

      //Serialize children
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing property-threshold");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_name()
    {
      return this.name;
    }
    public void Set_name(System.String value)
    {
      this.name = value;
      this.NotifyChange();
    }
    public System.Int32? Get_min_value_inclusive()
    {
      return this.min_value_inclusive;
    }
    public void Set_min_value_inclusive(System.Int32? value)
    {
      this.min_value_inclusive = value;
      this.NotifyChange();
    }
    public System.Int32? Get_max_value_inclusive()
    {
      return this.max_value_inclusive;
    }
    public void Set_max_value_inclusive(System.Int32? value)
    {
      this.max_value_inclusive = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      return null;
    }
  }
}